src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,220 @@
+    }
+
+    /// <summary>
+    /// Returns the path to the MEF composition cache file. Inputs used to determine the file name include:
+    /// 1) The given assembly paths
+    /// 2) The last write times of the given assembly paths
+    /// 3) The .NET runtime major version
+    /// </summary>
+    private static string GetCompositionCacheFilePath(string cacheDirectory, string catalogPrefix, ImmutableArray<string> assemblyPaths)
+    {
+        return Path.Combine(cacheDirectory, $"{catalogPrefix}.{ComputeAssemblyHash(assemblyPaths)}{CatalogSuffix}");
+
+        static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+        {
+            // Ensure AssemblyPaths are always in the same order.
+            assemblyPaths = assemblyPaths.Sort();
+
+            var hashContents = new StringBuilder();
+
+            // This should vary based on .NET runtime major version so that as some of our processes switch between our target
+            // .NET version and the user's selected SDK runtime version (which may be newer), the MEF cache is kept isolated.
+            // This can be important when the MEF catalog records full assembly names such as "System.Runtime, 8.0.0.0" yet
+            // we might be running on .NET 7 or .NET 8, depending on the particular session and user settings.
+            hashContents.Append(Environment.Version.Major);
+
+            foreach (var assemblyPath in assemblyPaths)
+            {
+                // Include assembly path in the hash so that changes to the set of included
+                // assemblies cause the composition to be rebuilt.
+                hashContents.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                hashContents.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
+            }
+
+            // Create base64 string of the hash.
+            var hashAsBase64String = Checksum.Create(hashContents.ToString()).ToString();
+
+            // Convert to filename safe base64 string.
+            return hashAsBase64String.Replace('+', '-').Replace('/', '_').TrimEnd('=');
+        }
+    }
+
+    private static async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, bool performCleanup, Action<string> logError, CancellationToken cancellationToken)
+    {
+        try
+        {
+            await Task.Yield().ConfigureAwait(false);
+
+            if (Path.GetDirectoryName(compositionCacheFile) is string directory)
+            {
+                var directoryInfo = Directory.CreateDirectory(directory);
+
+                if (performCleanup)
+                {
+                    // Delete any existing cached files.
+                    foreach (var fileInfo in directoryInfo.EnumerateFiles($"*{CatalogSuffix}"))
+                    {
+                        fileInfo.Delete();
```

Comment:
I assume you want a try/catch around this -- if this fails to cleanup you probably just want to just leave the other one there and assume it'll get cleaned up the next time.

Reply:
makes sense